Sources/SwiftParser/Specifiers.swift

```diff
@@ -415,7 +415,108 @@ extension RawAccessorEffectSpecifiersSyntax: RawEffectSpecifiersTrait {
       }
     }
   }
+}
+
+// This conformance exists solely to parse misplaced specifiers in parseFunctionReturnClause()
+extension RawDeinitEffectSpecifiersSyntax: RawEffectSpecifiersTrait {
+  enum MisspelledAsyncSpecifiers: TokenSpecSet {
+    case await
+    case reasync
+
+    init?(lexeme: Lexer.Lexeme) {
+      switch PrepareForKeywordMatch(lexeme) {
+      case TokenSpec(.await, allowAtStartOfLine: false): self = .await
+      case TokenSpec(.reasync): self = .reasync
+      default: return nil
+      }
+    }
+
+    var spec: TokenSpec {
+      switch self {
+      case .await: return TokenSpec(.await, allowAtStartOfLine: false)
+      case .reasync: return .keyword(.reasync)
+      }
+    }
+  }
+
+  enum CorrectAsyncTokenKinds: TokenSpecSet {
+    case async
+
+    init?(lexeme: Lexer.Lexeme) {
+      switch PrepareForKeywordMatch(lexeme) {
+      case TokenSpec(.async): self = .async
+      default: return nil
+      }
+    }
+
+    var spec: TokenSpec {
+      switch self {
+      case .async: return .keyword(.async)
+      }
+    }
+  }
+
+  enum MisspelledThrowsTokenKinds: TokenSpecSet {
+    case `rethrows`
+    case `try`
+    case `throw`
+    case `throws`
+
+    init?(lexeme: Lexer.Lexeme) {
+      switch PrepareForKeywordMatch(lexeme) {
+      case TokenSpec(.rethrows): self = .rethrows
+      case TokenSpec(.try, allowAtStartOfLine: false): self = .try
+      case TokenSpec(.throw, allowAtStartOfLine: false): self = .throw
+      case TokenSpec(.throws): self = .throws
+      default: return nil
+      }
+    }
+
+    var spec: TokenSpec {
+      switch self {
+      case .rethrows: return .keyword(.rethrows)
+      case .try: return TokenSpec(.try, allowAtStartOfLine: false)
+      case .throw: return TokenSpec(.throw, allowAtStartOfLine: false)
+      case .throws: return .keyword(.throws)
+      }
+    }
+  }
+
+  enum CorrectThrowsTokenKinds: TokenSpecSet {
+    // Uninhabited
 
+    init?(lexeme: Lexer.Lexeme) {
+      return nil
+    }
+
+    var spec: TokenSpec {
+      switch self {
+      }
+    }
+  }
+
+  var unexpectedBetweenAsyncSpecifierAndThrowsSpecifier: RawUnexpectedNodesSyntax? { unexpectedAfterAsyncSpecifier }
+  var throwsSpecifier: RawTokenSyntax? { nil }
+  var unexpectedAfterThrowsSpecifier: RawUnexpectedNodesSyntax? { nil }
+
+  init(
+    _ unexpectedBeforeAsyncSpecifier: RawUnexpectedNodesSyntax?,
+    asyncSpecifier: RawTokenSyntax?,
+    _ unexpectedBetweenAsyncSpecifierAndThrowsSpecifier: RawUnexpectedNodesSyntax?,
+    throwsSpecifier: RawTokenSyntax?,
+    _ unexpectedAfterThrowsSpecifier: RawUnexpectedNodesSyntax?,
+    arena: __shared SwiftSyntax.SyntaxArena
+  ) {
+    // Inserted missing throws is discarded
+    assert(throwsSpecifier?.isMissing ?? true)
+    assert(unexpectedAfterThrowsSpecifier == nil)
```

Comment:
This worries me a little bit because I’m not entirely sure that we really always end up with this being `nil` but can’t come up with a counter example from the top of my head. I would feel more comfortable if we just combine it with `unexpectedBetweenAsyncSpecifierAndThrowsSpecifier`.

```swift
RawUnexpectedNodesSyntax(combining: unexpectedBetweenAsyncSpecifierAndThrowsSpecifier, unexpectedAfterThrowsSpecifier, arena: arena),
```

Reply:
At this point we have only a single case (as noted in the extension comment), where we know that `unexpectedAfterThrowsSpecifier` cannot appear. Combining two groups of unexpected syntax looses information, which was affecting how `exchangeTokens()` behaves. Not sure if it still does after loosening the checks. But I would postpone combining or any other handling logic until other cases appear.

Reply:
I’m mostly thinking that the assumption that `unexpectedAfterThrowsSpecifier` is empty might change in the future. And especially the recovery cases are not hit as frequently in CI but more infrequently while code is being edited, so we might not actually notice that the assertion starts being violated. Thus, I’d rather be safe here. 

And I don’t see how `RawUnexpectedNodesSyntax(combining:)` can change the layout of the node if `unexpectedAfterThrowsSpecifier` is `nil` because it just grabs the elements of `unexpectedBetweenAsyncSpecifierAndThrowsSpecifier` and puts them in a new `RawUnexpectedNodesSyntax`.

Reply:
IMO, this entire conformance is hacky, and I would advise against introducing any new cases relying on it. I think even with `RawUnexpectedNodesSyntax(combining:)` implementation of the initialiser would be not robust enough to work in the general case. It is still not possible to pass `RawDeinitEffectSpecifiersSyntax` to `parseEffectSpecifiers()`.

Cleaner solution would be create a new protocol just for the `parseMisplacedEffectSpecifiers()`. I've pushed an updated. @ahoppen, WDYT?